File: chaingreen/consensus/block_rewards.py
from chaingreen.util.ints import uint32, uint64

# 1 Chaingreen coin = 1,000,000,000,000 = 1 trillion mio.
_mio_per_chaingreen = 1000000000000
_blocks_per_year = 1681920  # 32 * 6 * 24 * 365


def calculate_pool_reward(height: uint32) -> uint64:
    """
    Returns the pool reward at a certain block height. The pool earns 7/8 of the reward in each block. If the farmer
    is solo farming, they act as the pool, and therefore earn the entire block reward.
    These halving events will not be hit at the exact times
    (3 years, etc), due to fluctuations in difficulty. They will likely come early, if the network space and VDF
    rates increase continuously.
    """
    if height == 0:
        return uint64(0)  # genesis block has no reward
    elif height < 4 * _blocks_per_year:
        return uint64(int((7 / 8) * 500 * _mio_per_chaingreen))
    elif height < 8 * _blocks_per_year:
        return uint64(int((7 / 8) * 250 * _mio_per_chaingreen))
    elif height < 12 * _blocks_per_year:
        return uint64(int((7 / 8) * 125 * _mio_per_chaingreen))
    elif height < 16 * _blocks_per_year:
        return uint64(int((7 / 8) * 72.5 * _mio_per_chaingreen))
    elif height < 20 * _blocks_per_year:
        return uint64(int((7 / 8) * 31.25 * _mio_per_chaingreen))
    elif height < 24 * _blocks_per_year:
        return uint64(int((7 / 8) * 15.625 * _mio_per_chaingreen))
    elif height < 28 * _blocks_per_year:
        return uint64(int((7 / 8) * 7.8125 * _mio_per_chaingreen))
    else:
        return uint64(0)


def calculate_base_farmer_reward(height: uint32) -> uint64:
    """
    Returns the base farmer reward at a certain block height.
    The base fee reward is 1/8 of total block reward

    Returns the coinbase reward at a certain block height. These halving events will not be hit at the exact times
    (3 years, etc), due to fluctuations in difficulty. They will likely come early, if the network space and VDF
    rates increase continuously.
    """
    if height < 4 * _blocks_per_year:
        return uint64(int((1 / 8) * 500 * _mio_per_chaingreen))
    elif height < 8 * _blocks_per_year:
        return uint64(int((1 / 8) * 250 * _mio_per_chaingreen))
    elif height < 12 * _blocks_per_year:
        return uint64(int((1 / 8) * 125 * _mio_per_chaingreen))
    elif height < 16 * _blocks_per_year:
        return uint64(int((1 / 8) * 72.5 * _mio_per_chaingreen))
    elif height < 20 * _blocks_per_year:
        return uint64(int((1 / 8) * 31.25 * _mio_per_chaingreen))
    elif height < 24 * _blocks_per_year:
        return uint64(int((1 / 8) * 15.625 * _mio_per_chaingreen))
    elif height < 28 * _blocks_per_year:
        return uint64(int((1 / 8) * 7.8125 * _mio_per_chaingreen))
    else:
        return uint64(0)

File: chaingreen-gui-test/src/util/blockRewards.ts
import Big from 'big.js';

const MIO_PER_CHAINGREEN = Big(1000000000000);
const BLOCKS_PER_YEAR = 1681920;

export function calculatePoolReward(height: number): Big {
  if (height < 4 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(50).times(7 / 8);
  }
  if (height < 8 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(25).times(7 / 8);
  }
  if (height < 12 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(12.5).times(7 / 8);
  }
  if (height < 16 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(6.25).times(7 / 8);
  }
  if (height < 20 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(3.125).times(7 / 8);
  }
  if (height < 24 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(1.5625).times(7 / 8);
  }
  if (height < 28 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(0.78125).times(7 / 8);
  }

  return MIO_PER_CHAINGREEN.times(0).times(7 / 8);
}

export function calculateBaseFarmerReward(height: number): Big {
  if (height < 4 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(50).times(1 / 8);
  }
  if (height < 8 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(25).times(1 / 8);
  }
  if (height < 12 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(12.5).times(1 / 8);
  }
  if (height < 16 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(6.25).times(1 / 8);
  }
  if (height < 20 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(3.125).times(1 / 8);
  }
  if (height < 24 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(1.5625).times(1 / 8);
  }
  if (height < 28 * BLOCKS_PER_YEAR) {
    return MIO_PER_CHAINGREEN.times(0.78125).times(1 / 8);
  }

  return MIO_PER_CHAINGREEN.times(0).times(7 / 8);
}
